### PR TITLE
build-fhs-userenv-bubblewrap: Bind symlinks as symlinks in `/etc`

### LINKED ISSUE #224410: wine: can't see symlinked files created by FHS environment
### Describe the bug
When Wine is launched though an FHS environment, it can't see any symlinked file that has been generated by the FHS environment.

### Steps To Reproduce
An example:
You open Bottles (which uses an FHS environment that contains `alsa-plugins`), you launch a game with Wine (it doesn't matter if system Wine or a Wine downloaded with Bottles), and you see this error message:
```
ALSA lib conf.c:4098:(config_file_open) cannot access file /etc/alsa/conf.d/10-speexrate.conf
```

Let's investigate:
- You open a file manager through the same FHS environment, and the file is there (as a symlink).
- Then you try with `wine explorer`, and the file isn't there.

If you hackily copy-paste that file in that directory (and remove `alsa-plugins` from the FHS), then the file will appear in `wine explorer` and the error message won't appear anymore.

### Expected behavior
Symlinks created by the FHS environment should still be visible to Wine.
(Note that symlinks that exist normally and point to a valid location are visible. Directory symlinks are also visible.)

### Screenshots
image
The entire directory `/etc/alsa/conf.d/` has been replaced with new files by the FHS environment. All symlinks are gone in `wine explorer`.

### Metadata
```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.2.7-xanmod1, NixOS, 23.05 (Stoat), 23.05.20230331.f30990f`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.3`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by paveloom**
## Description of changes

Bind symbolic links in `/etc` via `--symlink`, too, as done below. This is a small extension to https://github.com/NixOS/nixpkgs/pull/183874. Fixes https://github.com/NixOS/nixpkgs/issues/224410.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SuperSamus** [comment]
@Atemu (sorry for your second ping on this)

**Comment by Atemu** [comment]
No need to feel sorry; I just haven't gotten around to reviewing it yet.

What would potential downsides of this approach be? Wouldn't this introduce a snapshot of partially stale state into the container?

**Comment by ParetoOptimalDev** [comment]
It's a long shot, but could this be a reason for why Apex Legends always has to recompile shaders on NixOS?

**Comment by Atemu** [comment]
I don't see how that would relate to this.

**Comment by hellwolf** [comment]
Fwiw, this patch also fixes a bug related to https://github.com/NixOS/nixpkgs/pull/279426

File: pkgs/build-support/build-fhsenv-bubblewrap/default.nix [hunk 1]
```diff
@@ -141,7 +141,11 @@ let
         if [[ $path == '/fonts' || $path == '/ssl' ]]; then
           continue
         fi
-        ro_mounts+=(--ro-bind "$i" "/etc$path")
+        if [[ -L $i ]]; then
+          symlinks+=(--symlink "$(${coreutils}/bin/readlink "$i")" "/etc$path")
```
Comment (hellwolf) [inline pkgs/build-support/build-fhsenv-bubblewrap/default.nix:145]: should it be `readlink -f` ?

File: pkgs/build-support/build-fhsenv-bubblewrap/default.nix [hunk 2]
```diff
@@ -141,7 +141,11 @@ let
...
+        else
+          ro_mounts+=(--ro-bind "$i" "/etc$path")
```
Comment (hellwolf) [inline pkgs/build-support/build-fhsenv-bubblewrap/default.nix:147]: is it possible that we always use `--symlink`, what would be the downside?

**Comment by hellwolf** [comment]
as an example, /etc/mtab is 0 byte (invalid with ro-bind) currently.

image

**Comment by paveloom** [comment]
Maintaining this PR is not in my interest anymore.

**Comment by TotalChaos05** [comment]
are there any workarounds for this (besides just using the unwrapped pkg)

**Comment by hellwolf** [comment]
It seems @Atemu is the maintainer of this, I could start a fresh PR on this one.

**Comment by Atemu** [comment]
Feel free, I can't promise a timely review though.

**Pull request closed without merging**